File: backend/lambda_functions/booking_request_lambda.js
const AWS = require("aws-sdk");
const uuid = require("uuid");
const crypto = require("crypto");

const sqs = new AWS.SQS();
const QUEUE_URL = "your_queue_url_here";

exports.handler = async (event, context) => {
  try {
    if (!event.body) {
      throw new Error("Missing 'body' in the event");
    }

    const bookingDetails = JSON.parse(event.body);

    const requiredFields = ["email", "room_id"];
    for (const field of requiredFields) {
      if (!bookingDetails[field]) {
        throw new Error(`Missing required field: ${field}`);
      }
    }

    // Generate a unique booking_id
    const bookingId = uuid.v4();

    // Construct the booking details with generated booking_id
    const bookingDetailsWithId = {
      booking_id: bookingId,
      email: bookingDetails.email,
      room_id: bookingDetails.room_id,
    };

    const messageGroupId = bookingDetails.room_id;
    const messageDeduplicationId = crypto
      .createHash("md5")
      .update(JSON.stringify(bookingDetailsWithId))
      .digest("hex");

    const params = {
      QueueUrl: QUEUE_URL,
      MessageBody: JSON.stringify(bookingDetailsWithId),
      MessageGroupId: messageGroupId,
      MessageDeduplicationId: messageDeduplicationId,
    };

    const response = await sqs.sendMessage(params).promise();

    // Add CORS headers to the response
    const headers = {
      "Access-Control-Allow-Origin": "*",
      "Access-Control-Allow-Headers": "Content-Type",
      "Access-Control-Allow-Methods": "POST",
    };

    return {
      statusCode: 200,
      headers: headers,
      body: JSON.stringify({
        message: "Booking details added to the queue",
        messageId: response.MessageId,
      }),
    };
  } catch (error) {
    return {
      statusCode: 500,
      headers: headers,
      body: JSON.stringify({
        error: "An error occurred",
        details: error.toString(),
      }),
    };
  }
};

File: backend/lambda_functions/booking_request_lambda.py
def lambda_handler(event, context):
    try:
        if 'body' not in event:
            raise ValueError("Missing 'body' in the event")
        
        booking_details = json.loads(event['body'])
        
        required_fields = ['email', 'room_id']
        for field in required_fields:
            if field not in booking_details:
                raise ValueError(f"Missing required field: {field}")

        # Generate a unique booking_id
        booking_id = str(uuid.uuid4())

        # Construct the booking details with generated booking_id
        booking_details_with_id = {
            'booking_id': booking_id,
            'email': booking_details['email'],
            'room_id': booking_details['room_id']
        }

        message_group_id = booking_details['room_id']
        message_deduplication_id = hashlib.md5(json.dumps(booking_details_with_id).encode('utf-8')).hexdigest()

        response = sqs.send_message(
            QueueUrl=QUEUE_URL,
            MessageBody=json.dumps(booking_details_with_id),
            MessageGroupId=message_group_id,
            MessageDeduplicationId=message_deduplication_id
        )
        
        # Add CORS headers to the response
        headers = {
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Headers': 'Content-Type',
            'Access-Control-Allow-Methods': 'POST'
        }
        
        return {
            'statusCode': 200,
            'headers': headers,
            'body': json.dumps({
                'message': 'Booking details added to the queue',
                'messageId': response['MessageId']
            })
        }
    except ValueError as ve:
        return {
            'statusCode': 400,
            'headers': headers,
            'body': json.dumps({
                'error': str(ve)
            })
        }
    except Exception as e:
        return {
            'statusCode': 500,
            'headers': headers,
            'body': json.dumps({
                'error': 'An error occurred',
                'details': str(e)
            })
        }